### PR TITLE
feat: Add .NET 6 targets & add support DispatcherQueue for WinUI

### PR DESCRIPTION
**Pull request by GabrielChouinardLetourneau**
GitHub Issue: #

## Proposed Changes

 - [ ] Bug fix
 - [x] Feature
 - [ ] Code style update (formatting)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build or CI related changes
 - [ ] Documentation content changes
 - [ ] Other, please describe:

## What is the current behavior?

There's no target for .NET 6 at the moment and WinUI isn't supported.

## What is the new behavior?

Add .NET 6 & WinUI support.

## Impact on version

- [ ] **Major** (Public API was modified.)
  - Public constructs (class, struct, delegate, enum, etc.) were removed or renamed.
  - Public members were removed or renamed.
  - Public method signatures were changed or renamed.
- [x] **Minor** (Public API was extended)
  - Public constructs, members, or overloads were added.
- [ ] **Patch** (Public API was unchanged.)
  - A bug in behavior was fixed.
  - Documentation was changed.

## Checklist

Please check that your PR fulfills the following requirements:

- [ ] Documentation has been added/updated.
- [ ] Automated Unit / Integration tests for the changes have been added/updated.
- [ ] Updated [BREAKING_CHANGES.md](../BREAKING_CHANGES.md) (if you introduced a breaking change).
- [ ] You conventional commit are aligned with the **Impact on version** section.

### CODE REVIEW

File: src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 	<PropertyGroup>
 		<TargetFrameworks>netstandard2.0</TargetFrameworks>
+		<LangVersion>10.0</LangVersion>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj:5]: Why would you need this?

File: src/DynamicMvvm.Uno.WinUI/Dispatchers/BatchingDispatcherQueueDispatcher.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+				_requests.Clear();
+			}
+
+			if (requests.Length == 0)
+			{
+				return;
+			}
+		
+
+			_dispatcherQueue.TryEnqueue(DispatcherQueuePriority.High, () =>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/Dispatchers/BatchingDispatcherQueueDispatcher.cs:89]: This is not functionally equivalent to the CoreDispatcher because it's not awaiting the completion of the task.

File: src/DynamicMvvm.Uno.WinUI/Dispatchers/DispatcherQueueDispatcher.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		/// <inheritdoc />
+		public async Task ExecuteOnDispatcher(CancellationToken ct, Action action)
+		{
+			if (ct.IsCancellationRequested)
+			{
+				this.Log().LogDebug($"Cancelled 'ExecuteOnDispatcher' because of the cancellation token.");
+				return;
+			}
+
+			_dispatcherQueue.TryEnqueue(DispatcherQueuePriority.High, () =>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/Dispatchers/DispatcherQueueDispatcher.cs:58]: This is not functionally equivalent to the CoreDispatcher because it's not awaiting the completion of the task.

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	</ItemGroup>
+	
+	<ItemGroup Condition="'$(TargetFramework)'=='net6.0-windows10.0.18362'">
+		<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.1.0" />
+		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" RuntimeFrameworkVersion="10.0.22000.24" />
+		<FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" TargetingPackVersion="10.0.22000.24" />
+	</ItemGroup>
+	
+	<ItemGroup Condition="'$(TargetFramework)'!='net6.0-windows10.0.18362'">
+		<Page Include="**\*.xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:50]: Why do we need this? We don't have any xaml files.

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		<Compile Update="**\*.xaml.cs">
+			<DependentUpon>%(Filename)</DependentUpon>
+		</Compile>
+	</ItemGroup>
+	
+	<ItemGroup>
+		<UpToDateCheckInput Include="**\*.xaml" Exclude="bin\**\*.xaml;obj\**\*.xaml" />
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:57]: same here

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

**Comment by jeanplevesque** [comment]
Do you really need 16 commits? Couldn't you squash them into 1?

File: src/DynamicMvvm.Abstractions/MvvmConfiguration.cs [hunk 6]
```diff
@@ -17,12 +17,12 @@ public static class DynamicMvvmConfiguration
 		/// </summary>
 		public static ILoggerFactory LoggerFactory { get; set; } = new NullLoggerFactory();
 
-		public static ILogger<T> Log<T>(this T _)
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Abstractions/MvvmConfiguration.cs]: Don't change this. It's a breaking change.

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 7]
```diff
@@ -0,0 +1,73 @@
+﻿<Project Sdk="MSBuild.Sdk.Extras/3.0.44">
+	<PropertyGroup>
+		<!-- Change the TargetFrameworks depending on which platform you are building on. This avoids errors as it is impossible to build UAP on OSX (MacOS) -->
+		<TargetFrameworks Condition="'$([MSBuild]::IsOsPlatform(OSX))'">netstandard2.0;net6.0-ios</TargetFrameworks>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:4]: @MatFillion Should we support the same platforms as Uno? (including macos, mac catalyst, etc.)

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+			<PackagePath>\</PackagePath>
+		</None>
+	</ItemGroup>
+	
+	<ItemGroup>
+		<PackageReference Include="MSBuild.Sdk.Extras" Version="3.0.44">
+			<PrivateAssets>all</PrivateAssets>
+			<IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
+		</PackageReference>
+		<PackageReference Include="Uno.WinUI" Version="4.4.13" />
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:40]: This should be the lowest major version.

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj (hunk 7 above)
Comment (Iguins) [reply src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:4]: There is a issue with Catalyst at the moment to build properly.
Maybe we could confirm with 17.3.3 if it's still happening

File: src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj (hunk 1 above)
Comment (Iguins) [reply src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj:5]: Could not compile with 17.2.X I believe with dotnet 6 :)

Comment (jeanplevesque) [reply src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj:5]: But that library is .net std 2.0 and doesn't use C# 10...

File: src/DynamicMvvm.Abstractions/MvvmConfiguration.cs (hunk 6 above)
Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Abstractions/MvvmConfiguration.cs]: This was an old change that was reverted in the last commit

Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Abstractions/MvvmConfiguration.cs]: It was put to public in [this commit](https://github.com/nventive/Chinook.DynamicMvvm/pull/71/commits/5fd6cd46aa1125acaea2e21618c4d2726db5d3f2)

File: src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj (hunk 1 above)
Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Abstractions/DynamicMvvm.Abstractions.csproj:5]: This was a mistake, this line is only necessary in the WinUI head

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj (hunk 8 above)
Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:40]: Do you mean the newest version 4.4.20 or lower than 4.4 (like 4.3)?

File: src/DynamicMvvm.Uno.WinUI/Dispatchers/BatchingDispatcherQueueDispatcher.cs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+				_requests.Clear();
+			}
+
+			if (requests.Length == 0)
+			{
+				return;
+			}
+
+
+			await DispatcherQueueExtensions.EnqueueAsync(_dispatcherQueue, () => ActivateRequests(requests), DispatcherQueuePriority.High);
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/Dispatchers/BatchingDispatcherQueueDispatcher.cs:90]: can't you use the extension method syntax?

File: src/DynamicMvvm.Uno/DynamicMvvm.Uno.csproj [hunk 10]
```diff
@@ -35,11 +35,17 @@
   </PropertyGroup>
 
   <ItemGroup Condition="'$(TargetFramework)'=='xamarinios10' or '$(TargetFramework)'=='monoandroid11.0' or '$(TargetFramework)'=='monoandroid12.0' or '$(TargetFramework)'=='netstandard2.0'">
-    <PackageReference Include="Uno.UI" Version="4.0.7" />
+	<PackageReference Include="MSBuild.Sdk.Extras" Version="3.0.44">
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno/DynamicMvvm.Uno.csproj:38]: Why do we need this new reference?

File: src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+using System.Runtime.CompilerServices;
+
+namespace MessageDialogService.Extensions
+{
+	internal static class DispatcherQueueExtensions
+	{
+		/// <summary>
+		/// Indicates whether or not <see cref="DispatcherQueue.HasThreadAccess"/> is available.
+		/// </summary>
+		private static readonly bool IsHasThreadAccessPropertyAvailable = ApiInformation.IsMethodPresent("Windows.System.DispatcherQueue", "HasThreadAccess");
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:19]: Are there cases when this can be false on the platforms we target?

File: src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+		/// <summary>
+		/// Invokes a given function on the target <see cref="DispatcherQueue"/> and returns a
+		/// <see cref="Task"/> that completes when the invocation of the function is completed.
+		/// </summary>
+		/// <param name="dispatcher">The target <see cref="DispatcherQueue"/> to invoke the code on.</param>
+		/// <param name="function">The <see cref="Action"/> to invoke.</param>
+		/// <param name="priority">The priority level for the function to invoke.</param>
+		/// <returns>A <see cref="Task"/> that completes when the invocation of <paramref name="function"/> is over.</returns>
+		/// <remarks>If the current thread has access to <paramref name="dispatcher"/>, <paramref name="function"/> will be invoked directly.</remarks>
+		public static Task EnqueueAsync(this DispatcherQueue dispatcher, Action function, DispatcherQueuePriority priority = DispatcherQueuePriority.Normal)
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:30]: I think we should rename this to RunAsync, or anything that's not "enqueue" because it doesn't sound like it's waiting for the execution.

File: src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+				{
+					return Task.FromException(e);
+				}
+			}
+
+			static Task TryEnqueueAsync(DispatcherQueue dispatcher, Action function, DispatcherQueuePriority priority)
+			{
+				var taskCompletionSource = new TaskCompletionSource<object?>();
+
+				if (!dispatcher.TryEnqueue(priority, () =>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:53]: I think it would be better to use a local function instead of a lambda here.

Comment (jeanplevesque) [reply src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:53]: We can probably remove the header of the file if we change the code significantly?
Hmmm not sure, might as well keep it.

File: src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs (hunk 12 above)
Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:30]: I'll rename it. I think they originally named it EnqueueAsync since DispatcherQueue uses TryEnqueue instead of RunAsync for CoreDispatcher. This extension is there to make sure it's async.

File: src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs (hunk 13 above)
Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:53]: I'll keep the header just since the inspiration came from their toolkit

File: src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs (hunk 11 above)
Comment (GabrielChouinardLetourneau) [reply src/DynamicMvvm.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs:19]: After checking a bit more, I think this bool isn't necessary here. It was used with the `DispatcherQueue` (`Windows.System`) that isn't used in the Windows App SDK, our `DispatcherQueue` is from `Microsoft.UI.Dispatching`. I will try something with the other implementation I made of this extension ([MessageDialogService](https://github.com/nventive/MessageDialogService/blob/master/src/MessageDialog.Uno.WinUI/Extensions/DispatcherQueueExtensions.cs)) to make sure removing this bool wouldn't break the `DispatcherQueue`.

**Review by Iguins — APPROVED** [review]

**Comment by jeanplevesque** [comment]
Don't forget to update the doc too :)

File: src/DynamicMvvm/DynamicMvvm.csproj [hunk 14]
```diff
@@ -2,6 +2,7 @@
 
 	<PropertyGroup>
 		<TargetFrameworks>netstandard2.0</TargetFrameworks>
+		<LangVersion>10.0</LangVersion>
```
Comment (jeanplevesque) [inline src/DynamicMvvm/DynamicMvvm.csproj:5]: Do you need this?

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+		<GenerateLibraryLayout>true</GenerateLibraryLayout>
+		<RootNamespace>Chinook.DynamicMvvm</RootNamespace>
+		<Authors>nventive</Authors>
+		<Company>nventive</Company>
+		<AssemblyName>Chinook.DynamicMvvm.Uno.WinUI</AssemblyName>
+		<PackageId>Chinook.DynamicMvvm.Uno.WinUI</PackageId>
+		<Description>Chinook.DynamicMvvm is a collection of extensible MVVM libraries for declarative ViewModels.</Description>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
+		<PackageReadmeFile>README.md</PackageReadmeFile>
+		<PackageTags>mvvm;xamarin;ios;android;uwp;chinook</PackageTags>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:17]: Those tags should be updated (for all csproj) now that winui and dotnet6 are available.

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: README.md [hunk 16]
```diff
@@ -102,6 +102,19 @@ public MainPage()
     };
 }
 ```
+#### Dispatcher Queue for WinUI
+When using the WinUI package, the Dispatcher given should be of type `DispatcherQueue`
```
Comment (jeanplevesque) [inline README.md:106]: I think you could rephrase this to explain that `CoreDispatcher` doesn't exist on WinUI and `DisptacherQueue` is the equivalent so the implementation names are different.

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+		<GenerateLibraryLayout>true</GenerateLibraryLayout>
+		<RootNamespace>Chinook.DynamicMvvm</RootNamespace>
+		<Authors>nventive</Authors>
+		<Company>nventive</Company>
+		<AssemblyName>Chinook.DynamicMvvm.Uno.WinUI</AssemblyName>
+		<PackageId>Chinook.DynamicMvvm.Uno.WinUI</PackageId>
+		<Description>Chinook.DynamicMvvm is a collection of extensible MVVM libraries for declarative ViewModels.</Description>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
+		<PackageReadmeFile>README.md</PackageReadmeFile>
+		<PackageTags>mvvm;ios;android;chinook;net-6.0;maui;net6.0-ios;winui;</PackageTags>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:17]: You can probably remove net6.0-ios because you already have ios and maui. (I checked online and net6.0-ios doesn't seem to be a popular tag, but rather a target framework.)

Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:17]: add the maui and winui tags in all other csproj too :)

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 18]
```diff
@@ -0,0 +1,55 @@
+﻿<Project Sdk="MSBuild.Sdk.Extras/3.0.44">
+	<PropertyGroup>
+		<!-- Change the TargetFrameworks depending on which platform you are building on. This avoids errors as it is impossible to build UAP on OSX (MacOS) -->
+		<TargetFrameworks Condition="'$([MSBuild]::IsOsPlatform(OSX))'">netstandard2.0;net6.0-ios;net6.0-maccatalyst;net6.0-macos</TargetFrameworks>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:4]: I don't think this condition is ever met because the pipeline only builds on windows...

File: src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+		<GenerateLibraryLayout>true</GenerateLibraryLayout>
+		<RootNamespace>Chinook.DynamicMvvm</RootNamespace>
+		<Authors>nventive</Authors>
+		<Company>nventive</Company>
+		<AssemblyName>Chinook.DynamicMvvm.Uno.WinUI</AssemblyName>
+		<PackageId>Chinook.DynamicMvvm.Uno.WinUI</PackageId>
+		<Description>Chinook.DynamicMvvm is a collection of extensible MVVM libraries for declarative ViewModels.</Description>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
+		<PackageReadmeFile>README.md</PackageReadmeFile>
+		<PackageTags>mvvm;ios;android;chinook;net-6.0;maui;winui;</PackageTags>
```
Comment (jeanplevesque) [inline src/DynamicMvvm.Uno.WinUI/DynamicMvvm.Uno.WinUI.csproj:15]: I think we can remove net-6.0. All nuget packages are for .net and the platforms are already interpreted by nuget.org

**Review by carlh98 — APPROVED** [review]
